### PR TITLE
Bug fixes and extensions for >4g dofs

### PR DESCRIPTION
**Pull request by kronbichler**
This PR introduces a series of bugfixes I had done for working on problems with more than 4g dofs.

In particular, the following changes were made:
- Utilities::MPI::Partitioner has been reworked to correctly distinguish between local indices (unsigned int) and global indices (types::global_dof_index).
- Iterators for DynamicSparsityPattern contained several bugs with respect to the index type. I discovered that it is still a bad idea to use an iterator from begin(row) to end(row) because end(row) jumps to the beginning of the next _relevant_ row. This is too bad on 10g unknowns. Thus, I removed the use of the iterators in the Trilinos sparse matrix and Trilinos sparsity pattern in favor of column_number.
- When working with more than 5000 processors, the approach to use a nonlocal graph that is filled with a dummy on each processor (in order to guarantee some Epetra-internal layout switching) is too bad. On >7000 cores my computations often got completely stuck because the communication model by Trilinos seems to end up in a black hole. Sometimes my cluster complained which allowed me to find the bug.
- parallel::distributed::Vector has been extended to the >4g case.
- DoFHandler::distribute_mg_dofs had a bug in computing the index sets for locally owned level dofs. (Another PR will be opened during the next days with fixes to MGTransferPrebuilt.)

### CODE REVIEW

File: include/deal.II/base/partitioner.h [hunk 1]
```diff
@@ -221,6 +221,17 @@ namespace Utilities
       bool is_compatible (const Partitioner &part) const;
 
       /**
+       * Checks whether the given partitioner is compatible with the
+       * partitioner used for this vector. Two partitioners are compatible if
+       * the have the same local size and the same ghost indices. They do not
```
Comment (Rombur) [inline include/deal.II/base/partitioner.h]: the have -> they have

File: include/deal.II/lac/parallel_vector.h [hunk 2]
```diff
@@ -971,6 +971,17 @@ namespace parallel
       bool
       partitioners_are_compatible (const Utilities::MPI::Partitioner &part) const;
 
+      /**
+       * Checks whether the given partitioner is compatible with the
+       * partitioner used for this vector. Two partitioners are compatible if
+       * the have the same local size and the same ghost indices. They do not
```
Comment (Rombur) [inline include/deal.II/lac/parallel_vector.h]: typo

File: source/base/partitioner.cc [hunk 3]
```diff
@@ -325,6 +333,40 @@ namespace Utilities
 
 
 
+    bool
+    Partitioner::is_compatible (const Partitioner &part) const
+    {
+      // is the partitioner points to the same memory location as the calling
```
Comment (Rombur) [inline source/base/partitioner.cc]: is -> if

File: source/base/partitioner.cc [hunk 4]
```diff
@@ -325,6 +333,40 @@ namespace Utilities
...
+              ghost_indices_data == part.ghost_indices_data);
+    }
+
+
+
+    bool
+    Partitioner::is_globally_compatible (const Partitioner &part) const
+    {
+      return Utilities::MPI::min(static_cast<int>(is_compatible(part)),
+                                 communicator) == 1;
```
Comment (Rombur) [inline source/base/partitioner.cc]: Why do check for the communicator? Don't you do it already in ```is_compatible```?

File: source/lac/trilinos_sparse_matrix.cc [hunk 5]
```diff
@@ -695,14 +707,11 @@ namespace TrilinosWrappers
...
-              // insert dummy element
-              TrilinosWrappers::types::int_type row =
-                nonlocal_graph->RowMap().MyGID(TrilinosWrappers::types::int_type(0));
-              nonlocal_graph->InsertGlobalIndices(row, 1, &row);
-            }
+          // must make sure the IndicesAreGlobal flag is set on all processors
+          // because some processors might not call InsertGlobalIndices (and
+          // we do not want to insert dummy indices on all processors for
+          // large-scale simulations due to the bad impact on performance)
+          nonlocal_graph->SetIndicesAreGlobal();
```
Comment (Rombur) [inline source/lac/trilinos_sparse_matrix.cc]: Isn't this already done by the constructor?

Comment (kronbichler) [inline source/lac/trilinos_sparse_matrix.cc]: Unfortunately not (at least I do not see how I could do it). The same code is present in Epetra_FECrsGraph, but there access is possible because it is derived from Epetra_CrsGraph, so I had to be a bit creative here. Not nice but works.

File: tests/mpi/parallel_vector_16.cc [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+// build a vector whose elements exceed the size of unsigned int in case of 64
+// bit indices. To avoid excessive memory consumption, let the vector start at
+// a number close to the maximum of unsigned int but extend past the last
+// index
```
Comment (Rombur) [inline tests/mpi/parallel_vector_16.cc]: What happens with this test when you only have 32bit indices?

File: source/base/partitioner.cc (hunk 4 above)
Comment (kronbichler) [inline source/base/partitioner.cc]: what do you mean here? MPI_Comm_compare is in is_compatible (local version). It is meant for the case where one vector is default-constructed (MPI_COMM_SELF) and the other is a distributed vector that neither owns nor ghosts vector elements for a given processor. Without this check, some processors would see a compatible partitioner here whereas it clearly is incompatible.

Comment (kronbichler) [inline source/base/partitioner.cc]: I should add that is_compatible is meant as a means to have lightweight checks without communication whereas the new is_globally_compatible function can be used to check things globally.

Comment (Rombur) [inline source/base/partitioner.cc]: OK that makes sense thanks.

File: tests/mpi/parallel_vector_16.cc (hunk 6 above)
Comment (kronbichler) [inline tests/mpi/parallel_vector_16.cc]: The test will fail early because I create an IndexSet with global size 4*42-39 but then I try to insert a range that starts close to the maximum size of unsigned int. The test is a bit artificial but I couldn't come up with a good test for the real problems I had because vectors would be at least 16G big and we can't do this on testing machines...

**Comment by kronbichler** [comment]
I made the minor typo fixes requested.

File: tests/mpi/parallel_vector_16.cc (hunk 6 above)
Comment (Rombur) [inline tests/mpi/parallel_vector_16.cc]: My bad, I didn't see that the test would not be run for 32bit.

File: tests/mpi/parallel_vector_16.cc [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+void test ()
+{
+  unsigned int myid = Utilities::MPI::this_mpi_process (MPI_COMM_WORLD);
+  unsigned int numproc = Utilities::MPI::n_mpi_processes (MPI_COMM_WORLD);
+
+  if (myid==0) deallog << "numproc=" << numproc << std::endl;
+
+  types::global_dof_index min_index =
+    types::global_dof_index(numbers::invalid_unsigned_int) - 39;
```
Comment (bangerth) [inline tests/mpi/parallel_vector_16.cc]: This may be fragile because it uses the fact that this number is a 32-bit integer. That's also not obvious.

Could you replace this by `0xffffffffU` instead?

**Comment by Rombur** [comment]
OK it's good for me.

File: tests/mpi/parallel_vector_16.cc (hunk 7 above)
Comment (kronbichler) [inline tests/mpi/parallel_vector_16.cc]: Sure.

File: tests/mpi/parallel_vector_16.cc (hunk 6 above)
Comment (bangerth) [inline tests/mpi/parallel_vector_16.cc]: Thanks for thinking of a way to test this -- I don't think anyone ran anything with more than 2^32 DoFs since Bruno and Kainan did the initial conversion several years ago. I'm not surprised that the code has bitrotted.

File: source/base/partitioner.cc [hunk 8]
```diff
@@ -161,7 +161,13 @@ namespace Utilities
         }
 
       std::vector<types::global_dof_index> first_index (n_procs+1);
-      first_index[0] = 0;
+      // Allow non-zero start index for the vector. send this data to all
+      // processors
+      first_index[0] = local_range_data.first;
+      MPI_Bcast(&first_index[0], sizeof(types::global_dof_index), MPI_BYTE,
+                0, communicator);
```
Comment (bangerth) [inline source/base/partitioner.cc]: I think you can do this in a more elegant way if you used
```
  MPI_Bcast (&first_index[0], 1, DEAL_II_DOF_INDEX_MPI_TYPE, 0, communicator);
```
This avoids interpreting the number as an array of bytes.

File: source/grid/grid_tools.cc [hunk 9]
```diff
@@ -2662,7 +2662,7 @@ namespace GridTools
     static inline std::bitset<3> lookup (const MATCH_T &)
     {
       // The 1D case is trivial
-      return 4; // [true ,false,false]
+      return 1; // [true ,false,false]
```
Comment (bangerth) [inline source/grid/grid_tools.cc]: I suppose this is an unrelated fix for something else. Do you happen to know how this worked before?

File: tests/mpi/parallel_vector_16.cc (hunk 6 above)
Comment (kronbichler) [inline tests/mpi/parallel_vector_16.cc]: Well, testing with more than 2^32 DoFs is not that simple - at least if you do some real computations. My test went beyond 2^32 first on 3584 cores with a multigrid solver, and I'd actually not use it on less than 10k cores for production runs.

Comment (bangerth) [inline tests/mpi/parallel_vector_16.cc]: Yes, I know it's not trivial. That's why I particularly appreciate the fact that you found a way to test some of the stuff on less than a few 1000 cores :-)

File: source/grid/grid_tools.cc (hunk 9 above)
Comment (kronbichler) [inline source/grid/grid_tools.cc]: Indeed, this should not have been there. I guess the reason is that there are neither tests nor instatiations for 1D. Well, there are also not too many cases possible in 1D. I will address that in a separate commit so @tamiko can have a look.

File: source/base/partitioner.cc (hunk 8 above)
Comment (kronbichler) [inline source/base/partitioner.cc]: That was taken from the code surrounding it that predates these MPI types. I will add a commit which does this throughout.

Comment (bangerth) [inline source/base/partitioner.cc]: OK, thanks. I have no other comments -- since @Rombur seemed ok with it, feel free to merge whenever you're happy with the current state.
